File: tests/f_api/odc_encode_custom.f90
! To build this program, please make sure to first compile odc Fortran module,
! and then reference linked libraries:
!
!     gfortran -c ../../src/odc/api/odc.f90
!     gfortran -lodccore -lfodc -o odc-fortran-encode-custom odc_encode_custom.f90

program odc_ls
    use, intrinsic :: iso_c_binding, only: c_loc,c_null_char
    use odc
    implicit none

    character(255) :: path

    integer(8), parameter :: nrows = 20
    character(8), target :: data1(nrows)
    integer(8), target :: data2(nrows)
    character(8), target :: data3(nrows)
    character(16), target :: data4(nrows)
    real(8), target :: data5(nrows)
    integer(8), target :: data6(nrows)
    real(8), target :: data7(nrows)
    integer(8), target :: data8(nrows)

    type(odc_encoder) :: encoder
    character(10), parameter :: property_key = 'encoded_by'
    character(11), parameter :: property_value = 'odc_example'
    integer, target :: outunit
    integer(8), target :: bytes_written

    if (command_argument_count() /= 1) then
        call usage()
        stop 1
    end if

    ! Get output path from command argument
    call get_command_argument(1, path)

    ! Initialise API and set treatment of integers as longs
    call check_call(odc_initialise_api(), 'initialising api')
    call check_call(odc_integer_behaviour(ODC_INTEGERS_AS_LONGS), 'setting integer behaviour to longs')

    ! Set up the allocated array with scratch data
    call create_scratch_data(data1, data2, data3, data4, data5, data6, data7, nrows)

    ! Initialise encoder
    call check_call(encoder%initialise(), 'initialising encoder')

    ! Set number of rows to allocate in the encoder
    call check_call(encoder%set_row_count(nrows), 'setting number of rows')

    ! Define all column names and their types
    call check_call(encoder%add_column('expver', ODC_STRING), 'adding expver column')
    call check_call(encoder%add_column('date@hdr', ODC_INTEGER), 'adding date@hdr column')
    call check_call(encoder%add_column('statid@hdr', ODC_STRING), 'adding statid@hdr column')
    call check_call(encoder%add_column('wigos@hdr', ODC_STRING), 'adding wigos@hdr column')
    call check_call(encoder%add_column('obsvalue@body', ODC_REAL), 'adding obsvalue@bod column')
    call check_call(encoder%add_column('integer_missing', ODC_INTEGER), 'adding integer_missing column')
    call check_call(encoder%add_column('double_missing', ODC_REAL), 'adding double_missing column')
    call check_call(encoder%add_column('bitfield_column', ODC_BITFIELD), 'adding bitfield_column')

    ! Column `wigos@hdr` is a 16-byte string column
    call check_call(encoder%column_set_data_size(4, element_size_doubles=2), 'setting column data size')

    ! Column `bitfield_column` is an integer with 4 bitfield values in it
    call check_call(encoder%column_add_bitfield(8, "flag_a", 1), 'adding flag_a bitfield')
    call check_call(encoder%column_add_bitfield(8, "flag_b", 2), 'adding flag_b bitfield')
    call check_call(encoder%column_add_bitfield(8, "flag_c", 3), 'adding flag_c bitfield')
    call check_call(encoder%column_add_bitfield(8, "flag_d", 1), 'adding flag_d bitfield')

    ! Set a custom data layout and data array for each column
    call check_call(encoder%column_set_data_array(1, 8, stride=8, data=c_loc(data1)), 'setting expver array')
    call check_call(encoder%column_set_data_array(2, 8, stride=8, data=c_loc(data2)), 'setting date array')
    call check_call(encoder%column_set_data_array(3, 8, stride=8, data=c_loc(data3)), 'setting statid array')
    call check_call(encoder%column_set_data_array(4, 16, stride=16, data=c_loc(data4)), 'setting wigos array')
    call check_call(encoder%column_set_data_array(5, 8, stride=8, data=c_loc(data5)), 'setting obsvalue array')
    call check_call(encoder%column_set_data_array(6, 8, stride=8, data=c_loc(data6)), 'setting integer_missing array')
    call check_call(encoder%column_set_data_array(7, 8, stride=8, data=c_loc(data7)), 'setting double_missing array')
    call check_call(encoder%column_set_data_array(8, 8, stride=8, data=c_loc(data8)), 'setting bitfield_column array')

    ! Add some key/value metadata to the frame
    call check_call(encoder%add_property(property_key, property_value), 'adding property')

    ! Encode ODB-2 into a file
    open(newunit=outunit, file=path, access='stream', form='unformatted', status='replace')
    call check_call(encoder%encode(outunit, bytes_written), 'encoding data')
    close(outunit)

    ! Deallocate memory used up by the encoder
    call check_call(encoder%free(), 'cleaning up encoder')

    write(6, '(a,i0,a,a)') 'Written ', nrows, ' rows to ', trim(path)

contains

    subroutine check_call(err, desc)
        integer, intent(in) :: err
        character(*), intent(in) :: desc

        if (err /= ODC_SUCCESS) then
            write(7, *) '**** An error occurred in ODC library'
            write(7, *) 'Description: ', desc
            write(7, *) 'Error: ', odc_error_string(err)
            stop 1
        end if
    end subroutine

    subroutine usage()
        write(6, *) 'Usage:'
        write(6, *) '    odc-fortran-encode-custom <odb2 output file>'
    end subroutine

    subroutine cycle_ints(output, input)
        integer(8), intent(in) :: input(:)
        integer(8), intent(out) :: output(:)
        integer :: i

        do i = 1, size(output, 1)
            output(i) = input(mod(i - 1, size(input, 1)) + 1)
        end do
    end subroutine

    subroutine cycle_reals(output, input)
        real(8), intent(in) :: input(:)
        real(8), intent(out) :: output(:)
        integer :: i

        do i = 1, size(output, 1)
            output(i) = input(mod(i - 1, size(input, 1)) + 1)
        end do
    end subroutine

    subroutine create_scratch_data(data1, data2, data3, data4, data5, data6, data7, nrows)
        character(8), intent(out) :: data1(:)
        integer(8), intent(out) :: data2(:)
        character(8), intent(out):: data3(:)
        character(16), intent(out) :: data4(:)
        real(8), intent(out) :: data5(:)
        integer(8), intent(out) :: data6(:)
        real(8), intent(out) :: data7(:)
        integer(8), intent(in) :: nrows

        character(4) :: expver_str = 'xxxx'
        character(8) :: date_str
        integer(8) :: date
        character(6) :: statid_str
        character(16) :: wigos_str
        integer(8) :: missing_integer
        integer(8) :: integer_pool(3)
        integer(8), target :: missing_integers(nrows)
        real(8) :: missing_double
        real(8) :: double_pool(3)
        real(8), target :: missing_doubles(nrows)
        integer(8) :: bitfield_pool(3)
        integer(8), target :: bitfield_values(nrows)
        integer(8) :: i

        ! Prepare the current date as an integer
        call date_and_time(date=date_str)
        read(date_str, *) date

        ! Prepare the list of integer values, including the missing value
        call check_call(odc_missing_integer(missing_integer), 'getting missing integer value')
        integer_pool = (/ int(1234, 8), int(4321, 8), missing_integer /)
        call cycle_ints(missing_integers, integer_pool)

        ! Prepare the list of double values, including the missing value
        call check_call(odc_missing_double(missing_double), 'getting missing double value')
        double_pool = (/ real(12.34, 8), real(43.21, 8), missing_double /)
        call cycle_reals(missing_doubles, double_pool)

        ! Prepare the list of bitfield values
        bitfield_pool = (/ int(b'00000001'), int(b'00001011'), int(b'01101011') /)
        call cycle_ints(bitfield_values, bitfield_pool)

        ! Fill in the passed data arrays with scratch values
        do i = 1, nrows
            data1(i) = expver_str // c_null_char  ! expver
            data2(i) = date  ! date@hdr

            write(statid_str, '(a,i0.2)') 'stat', i - 1
            data3(i) = statid_str // c_null_char  ! statid@hdr

            write(wigos_str, '(a,i0.2,a)') '0-12345-0-678', i - 1, '' // c_null_char
            data4(i) = wigos_str  ! wigos@hdr

            data5(i) = 12.3456 * real(i - 1)  ! obsvalue@body
            data6(i) = missing_integers(i)  ! integer_missing
            data7(i) = missing_doubles(i)  ! double_missing
            data8(i) = bitfield_values(i)  ! bitfield_column
        end do
    end subroutine

end program

File: tests/c_api/odc_encode_custom.c
/**
 * To build this program, please make sure to reference linked library:
 *
 *     gcc -lodccore -o odc-c-encode-custom odc_encode_custom.c
 */

#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>

#include "odc/api/odc.h"

// Bitfield constants
#define Ob00000001 1
#define Ob00001011 11
#define Ob01101011 107

#define CHECK_RESULT(x)                                                                           \
    do {                                                                                          \
        int rc = (x);                                                                             \
        if (rc != ODC_SUCCESS) {                                                                  \
            fprintf(stderr, "Error calling odc function \"%s\": %s\n", #x, odc_error_string(rc)); \
            exit(1);                                                                              \
        }                                                                                         \
    } while (false);

void usage() {
    fprintf(stderr, "Usage:\n");
    fprintf(stderr, "    odc-c-encode-custom <odb2 output file>\n\n");
}

void cycle_longs(long* list, int size, long* pool, int pool_size) {
    int index = 0;
    int i;

    for (i = 0; i < size; i++) {
        if (index == pool_size)
            index = 0;
        list[i] = pool[index];
        index++;
    }
}

void cycle_doubles(double* list, int size, double* pool, int pool_size) {
    int index = 0;
    int i;

    for (i = 0; i < size; i++) {
        if (index == pool_size)
            index = 0;
        list[i] = pool[index];
        index++;
    }
}

void create_scratch_data(long nrows, char data0[][8], int64_t data1[], char data2[][8], char data3[][16],
                         double data4[], int64_t data5[], double data6[], int64_t data7[]) {

    // Prepare the current date as an integer

    time_t rawtime;
    time(&rawtime);

    struct tm* timeinfo;
    timeinfo = localtime(&rawtime);

    int64_t date = 10000 * (timeinfo->tm_year + 1900) + 100 * (timeinfo->tm_mon + 1) + timeinfo->tm_mday;

    // Prepare the list of integer values, including the missing value

    long missing_integer;
    CHECK_RESULT(odc_missing_integer(&missing_integer));

    long integer_pool[]   = {1234, 4321, missing_integer};
    int integer_pool_size = sizeof(integer_pool) / sizeof(integer_pool[0]);

    long missing_integers[nrows];
    cycle_longs(missing_integers, nrows, integer_pool, integer_pool_size);

    // Prepare the list of double values, including the missing value

    double missing_double;
    CHECK_RESULT(odc_missing_double(&missing_double));

    double double_pool[] = {12.34, 43.21, missing_double};
    int double_pool_size = sizeof(double_pool) / sizeof(double_pool[0]);

    double missing_doubles[nrows];
    cycle_doubles(missing_doubles, nrows, double_pool, double_pool_size);

    // Prepare the list of bitfield values

    long bitfield_pool[]   = {Ob00000001, Ob00001011, Ob01101011};
    int bitfield_pool_size = sizeof(bitfield_pool) / sizeof(bitfield_pool[0]);

    long bitfield_values[nrows];
    cycle_longs(bitfield_values, nrows, bitfield_pool, bitfield_pool_size);

    int i;

    // Fill in the passed data arrays with scratch values
    for (i = 0; i < nrows; i++) {
        snprintf(data0[i], 8, "xxxx");                   // expver
        data1[i] = date;                                 // date@hdr
        snprintf(data2[i], 7, "stat%02d", i);            // statid@hdr
        snprintf(data3[i], 16, "0-12345-0-678%02d", i);  // wigos@hdr
        data4[i] = 12.3456 * i;                          // obsvalue@body
        data5[i] = missing_integers[i];                  // integer_missing
        data6[i] = missing_doubles[i];                   // double_missing
        data7[i] = bitfield_values[i];                   // bitfield_column
    }
}

int main(int argc, char* argv[]) {
    if (argc != 2) {
        usage();
        return 1;
    }

    // Get output path from command argument
    char* path = argv[1];

    // Initialise API and set treatment of integers as longs
    CHECK_RESULT(odc_initialise_api());
    CHECK_RESULT(odc_integer_behaviour(ODC_INTEGERS_AS_LONGS));

    // Define row count
    const long nrows = 20;

    // Allocate data array for each column
    char data0[nrows][8];
    int64_t data1[nrows];
    char data2[nrows][8];
    char data3[nrows][16];
    double data4[nrows];
    int64_t data5[nrows];
    double data6[nrows];
    int64_t data7[nrows];

    // Set up the allocated array with scratch data
    create_scratch_data(nrows, data0, data1, data2, data3, data4, data5, data6, data7);

    // Initialise encoder
    odc_encoder_t* encoder = NULL;
    CHECK_RESULT(odc_new_encoder(&encoder));

    // Set number of rows to allocate in the encoder
    CHECK_RESULT(odc_encoder_set_row_count(encoder, nrows));

    // Define all column names and their types
    CHECK_RESULT(odc_encoder_add_column(encoder, "expver", ODC_STRING));
    CHECK_RESULT(odc_encoder_add_column(encoder, "date@hdr", ODC_INTEGER));
    CHECK_RESULT(odc_encoder_add_column(encoder, "statid@hdr", ODC_STRING));
    CHECK_RESULT(odc_encoder_add_column(encoder, "wigos@hdr", ODC_STRING));
    CHECK_RESULT(odc_encoder_add_column(encoder, "obsvalue@body", ODC_REAL));
    CHECK_RESULT(odc_encoder_add_column(encoder, "integer_missing", ODC_INTEGER));
    CHECK_RESULT(odc_encoder_add_column(encoder, "double_missing", ODC_REAL));
    CHECK_RESULT(odc_encoder_add_column(encoder, "bitfield_column", ODC_BITFIELD));

    // Column `wigos@hdr` is a 16-byte string column
    CHECK_RESULT(odc_encoder_column_set_data_size(encoder, 3, 16));

    // Column `bitfield_column` is an integer with 4 bitfield values in it
    CHECK_RESULT(odc_encoder_column_add_bitfield(encoder, 7, "flag_a", 1));
    CHECK_RESULT(odc_encoder_column_add_bitfield(encoder, 7, "flag_b", 2));
    CHECK_RESULT(odc_encoder_column_add_bitfield(encoder, 7, "flag_c", 3));
    CHECK_RESULT(odc_encoder_column_add_bitfield(encoder, 7, "flag_d", 1));

    // Set a custom data layout and data array for each column
    CHECK_RESULT(odc_encoder_column_set_data_array(encoder, 0, 8, 8, data0));
    CHECK_RESULT(odc_encoder_column_set_data_array(encoder, 1, sizeof(int64_t), sizeof(int64_t), data1));
    CHECK_RESULT(odc_encoder_column_set_data_array(encoder, 2, 8, 8, data2));
    CHECK_RESULT(odc_encoder_column_set_data_array(encoder, 3, 16, 16, data3));
    CHECK_RESULT(odc_encoder_column_set_data_array(encoder, 4, sizeof(double), sizeof(double), data4));
    CHECK_RESULT(odc_encoder_column_set_data_array(encoder, 5, sizeof(int64_t), sizeof(int64_t), data5));
    CHECK_RESULT(odc_encoder_column_set_data_array(encoder, 6, sizeof(double), sizeof(double), data6));
    CHECK_RESULT(odc_encoder_column_set_data_array(encoder, 7, sizeof(int64_t), sizeof(int64_t), data7));

    const char* property_key   = "encoded_by";
    const char* property_value = "odc_example";

    // Add some key/value metadata to the frame
    CHECK_RESULT(odc_encoder_add_property(encoder, property_key, property_value));

    int file_descriptor = open(path, O_CREAT | O_TRUNC | O_WRONLY, 0666);
    long size;

    // Encode ODB-2 into an already open file descriptor
    CHECK_RESULT(odc_encode_to_file_descriptor(encoder, file_descriptor, &size));

    close(file_descriptor);

    // Deallocate memory used up by the encoder
    CHECK_RESULT(odc_free_encoder(encoder));

    fprintf(stdout, "Written %ld rows to %s\n", nrows, path);

    return 0;
}
